### PR TITLE
Rewrite the AST Bridge.

### PR DESCRIPTION
**Pull request by schweitzpgi**
This is a major rewrite of the AST bridge code to make it uniformly use post-order traversal to convert the C++ AST to MLIR. It replaces the ad hoc, on-demand bridge that recursed over multiple instances of itself, etc. The new structure should be less unweildy and hopefully better suited to adding new features to the compiler as the project grows.

At the top-level this uses clang's recursive AST visitor, which is itself somewhat problematic. Clang's visitor is tuned for printing the AST like a mirror of the user's source code input. That means some of the necessary nodes in the AST are skipped and some of the semantics-free syntax is visited. Some sawing and sanding to smooth these rough edges is required.

The new overloads of the visitor make heavy use of Traverse* methods for choreography of what parts of the AST must be visited, what parts should not be visited, and setting up MLIR data structures prior to these visits. Specific node types generate code (or types) typically in the overloaded Visit* methods. Visiting types uses a Type stack and visiting code uses a Value stack. When generating IR, each Value carries its own Type and the type information is not duplicated on both stacks.

Some high-level structures, almost exclusively from the std and cudaq namespaces, are lowered to high-level ops and types in Quake. These substitutions have been consolidated somewhat but could be moreso. These substitutions add complexity to the bridge by having to check for the cases where the substitutions are required for correct IR.

The bridge generates more unneeded code now. To compensate, the cudaq-quake tool runs the canonicalizer. It should run symbol DCE, but there is still an outstanding bug, so that is disabled for the moment.

### CODE REVIEW

File: lib/Frontend/nvqpp/ConvertDecl.cpp [hunk 1]
```diff
@@ -153,66 +120,351 @@ QuakeBridgeVisitor::getOrAddFunc(Location loc, StringRef funcName,
...
+        if (size != static_cast<std::int64_t>(std::dynamic_extent))
+          return pushType(quake::VeqType::get(ctx, size));
+      }
+      return pushType(quake::VeqType::getUnsized(ctx));
+    }
+    // qvector<LEVEL>, qview<LEVEL>
+    if (name.equals("qvector") || name.equals("qview"))
+      return pushType(quake::VeqType::getUnsized(ctx));
+    if (name.equals("ctrl") /*|| name.equals("adj") || name.equals("base")*/)
+      return pushType({});
```
Comment (amccaskey) [inline lib/Frontend/nvqpp/ConvertDecl.cpp:170]: Will this cause issues for anything popping off the last type?

File: tools/cudaq-quake/cudaq-quake.cpp [hunk 2]
```diff
@@ -448,8 +453,15 @@ int main(int argc, char **argv) {
     }
 
     // Running the verifier to make it easier to track down errors.
+    // The canonicalizer and symbol DCE passes cleanup unused artifacts
+    // generated by the bridge.
     mlir::PassManager pm(&context);
     pm.addPass(std::make_unique<cudaq::VerifierPass>());
+    if (!noSimplify) {
+      pm.addPass(mlir::createCanonicalizerPass());
+      // FIXME: Workaround bug in regenerating the CPU side kernel entry.
```
Comment (amccaskey) [inline tools/cudaq-quake/cudaq-quake.cpp:462]: What exactly is the bug here? We haven't generated the new entrypoint function yet.

**Comment by amccaskey** [comment]
Double-check me, but when I run `cudaq-quake` with this branch, I see that the quantum operation functions are added as declarations to the module
```cpp
#include "cudaq.h"

__qpu__ void test() {
  cudaq::qubit q;
  ry(2.2, q);
}
```
```mlir
module attributes {quake.mangled_name_map = {__nvqpp__mlirgen__function_test._Z4testv = "_Z4testv"}} {
  func.func @__nvqpp__mlirgen__function_test._Z4testv() attributes {"cudaq-entrypoint", "cudaq-kernel"} {
    %cst = arith.constant 2.200000e+00 : f64
    %0 = quake.alloca !quake.ref
    quake.ry (%cst) %0 : (f64, !quake.ref) -> ()
    return
  }
  func.func private @_ZN5cudaq2ryINS_4baseEdJNS_5quditILm2EEEEEEvT0_DpRT1_(f64, !quake.ref)
  func.func private @_Z4testv()
}
```

The code is correct, but we don't want those functions still lying around.

File: lib/Frontend/nvqpp/ConvertDecl.cpp [hunk 3]
```diff
@@ -153,66 +120,351 @@ QuakeBridgeVisitor::getOrAddFunc(Location loc, StringRef funcName,
...
+    return false;
+  auto name = ident->getName();
+  auto *ctx = builder.getContext();
+  if (isInNamespace(x, "cudaq")) {
+    // Types from the `cudaq` namespace.
+    // A qubit is a qudit<LEVEL=2>.
+    if (name.equals("qudit") || name.equals("qubit"))
+      return pushType(quake::RefType::get(ctx));
+    // qreg<SIZE,LEVEL>, qarray<SIZE,LEVEL>, qspan<SIZE,LEVEL>
+    if (name.equals("qspan") || name.equals("qreg") || name.equals("qarray")) {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Frontend/nvqpp/ConvertDecl.cpp:152]: Well, there is some support. I didn't kick the tires very hard.

File: tools/cudaq-quake/cudaq-quake.cpp (hunk 2 above)
Comment (schweitzpgi) [reply tools/cudaq-quake/cudaq-quake.cpp:462]: GenKernelExecution is leaning on the CPU side (C++) entry points being declared by the bridge. There is code (actually the same routine) to figure it out if the declaration is missing, but it guesses wrong in 2 of the regression tests.

It is on my list of things to figure out in a follow-up.

File: lib/Frontend/nvqpp/ConvertDecl.cpp (hunk 1 above)
Comment (schweitzpgi) [reply lib/Frontend/nvqpp/ConvertDecl.cpp:170]: This is a NOP to meet the invariant here. If the type is intercepted (meaning the bridge is filtering out this C++ type and replacing it in whole with a Quake type, ignoring the C++ definition), then the invariant is that a `mlir::Type` is pushed on the top of the stack. For `cudaq::ctrl` that Quake type is very uninteresting and technically-speaking erased. So this null Type will be popped and the code will generate an mlir attribute.

On the other hand, I think a case could be made that nothing should be pushed on the type stack here, as the other two aren't doing so.

Bottom line: I need to investigate this more.

**Review by bmhowe23 — COMMENTED** [review]
Just a few comments below. And for an extra warm fuzzy, I ran all of the `test/NVQPP/integration/*.cpp` tests with `--target quantinuum --emulate` and CUDAQ_DUMP_JIT_IR=1, and I compared all the results back to the current main branch. The LLVM IR output was a perfect match for all 13 test cases. 👍

File: include/cudaq/Frontend/nvqpp/ASTBridge.h [hunk 4]
```diff
@@ -315,15 +392,8 @@ class QuakeBridgeVisitor
...
-  bool shouldTraversePostOrder() {
-    if (typeMode)
-      return true;
-    bool result = postOrderTraversal;
-    if (!result)
-      postOrderTraversal = true;
-    return result;
-  }
+  // The AST is visited in postorder.
+  bool shouldTraversePostOrder() { return true; }
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply include/cudaq/Frontend/nvqpp/ASTBridge.h:396]: Yes. It is used everywhere, actually. It's a callback in the recursive AST visitor code.

File: include/cudaq/Frontend/nvqpp/ASTBridge.h [hunk 5]
```diff
@@ -231,43 +282,79 @@ class QuakeBridgeVisitor
...
+    return TraverseType(t->desugar());
+  }
+  bool TraverseTypeOfExprType(clang::TypeOfExprType *t) {
+    // Do not visit the expression as it is has no semantics other than for
+    // inferring a type.
+    return true;
+  }
+  bool TraverseNestedNameSpecifier(clang::NestedNameSpecifier *s) {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply include/cudaq/Frontend/nvqpp/ASTBridge.h:340]: Again these are callbacks in the CRTP code.

File: tools/cudaq-quake/cudaq-quake.cpp [hunk 6]
```diff
@@ -73,6 +73,11 @@ static cl::opt<bool>
     verifyMode("verify", cl::desc("Run in diagnostic verification mode."),
                cl::init(false));
 
+static cl::opt<bool>
+    noSimplify("no-simplify",
+               cl::desc("Run passes to simplify the code after the bridge."),
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply tools/cudaq-quake/cudaq-quake.cpp:78]: Thanks, Ben. Good catch.

**Review by amccaskey — APPROVED** [review]
+1
